Add opt-in timeout to tests

rymaju opened the pull request:
Based on the conversation in  #22 

- Adds a parameter `timeout-in-seconds`, which is either 
  - a number that determines how long to run `generate-results`
  - `#f` (default), representing no timeout
- `call/timeout`, a wrapper that calls a procedure and returns the result -- sending a break after `timeout-in-seconds`.
  - This behavior is pretty useful for `fold-test-cases`: when you send a break, all pending tests fail!

I made this opt-in so that this doesn't add a timeout to existing test suites.

In your `grade-xxx.rkt`, you might write something like:

```racket
#lang racket

(require rackunit)
...

(define-test-suite sq-tests
  (test-equal? "0" ((lambda (x) (x x)) (lambda (x) (x x))) 0) ; infinite loop!
  ...)

(parameterize ([timeout-in-seconds 5]) ; parameterize timeout here
  (generate-results sq-tests))
```

In the generated results, timed out tests show up exactly like test-errors:
```
Execution error in test named «suite:test-name»
```

## Verification

Ran with modified `grade-sq.rkt`s, parameterized 

### Infinite loop + Error + Ok
Code:
```rkt
(define-test-suite sq-tests
  (test-equal? "0" ((lambda (x) (x x)) (lambda (x) (x x))) 0)
  (test-equal? "1" (sq 1) 0)
  (test-equal? "-1" (sq -1) 1))

(parameterize ([timeout-in-seconds 5])
  (generate-results sq-tests))
```

Results + Time:
```
...
docker run -ti \
		-v `pwd`/tests/sq/s1:/autograder/submission \
		-v `pwd`/autograder/results:/autograder/results \
		shriramk/gradescope-racket-sq /autograder/run_autograder
printf "\n\n"


cat autograder/results/results.json
{"score":"33.333333333333336","tests":[{"output":"Execution error in test named «sq-tests:0»"},{"output":"Incorrect answer from test named «sq-tests:1»"}]}printf "\n\n"


make grade a=sq s=sq/s1  0.24s user 0.37s system 6% cpu 9.563 total
```

### Error + Ok
Code:
```rkt
(define-test-suite sq-tests
  ;(test-equal? "0" ((lambda (x) (x x)) (lambda (x) (x x))) 0)
  (test-equal? "1" (sq 1) 0)
  (test-equal? "-1" (sq -1) 1))

(parameterize ([timeout-in-seconds 5])
  (generate-results sq-tests))
```

Results + Time:
```
...
docker run -ti \
		-v `pwd`/tests/sq/s1:/autograder/submission \
		-v `pwd`/autograder/results:/autograder/results \
		shriramk/gradescope-racket-sq /autograder/run_autograder
printf "\n\n"


cat autograder/results/results.json
{"score":"50.0","tests":[{"output":"Incorrect answer from test named «sq-tests:1»"}]}printf "\n\n"


make grade a=sq s=sq/s1  0.28s user 0.47s system 19% cpu 3.763 total
```

@jasonhemann

rymaju committed:
feat: timeout tests

- add call/timeout, a function that takes a thunk and returns the result, breaking after a timeout (default 1s)
- wrap fold-test-results in call/timeout, due to the behavior of run-test-case, this will cause incomplete tests to break and error rather than hanging

Commit:
🐛 fix: handle no hanging tests in call/timeout

Previously, call/timeout failed with a user break when no all tests were
passing or errored immediately.

Now, we check if there are immediate results before breaking the thread.

Other improvements:
- parameterized timeout-in-seconds in call/timeout
- use async channels so that suites without hanging tests pass/fail
immediately rather than waiting on timeout

jasonhemann commented:
I'm still running into issues trying to timeout with a small `sq`-based hanging impl. 

```
;; tests/sq/s10/defs.rkt
#lang racket/base

(define (sq n)
  (sq n))
```

```
;; changing grade-sq.rkt to 
... 

(parameterize ([timeout-in-seconds 5])
  (generate-results sq-tests))
```

hangs when I execute the makefile on my local system:

```
 make a=sq s=sq/s10
cp grade-sq.rkt grade.rkt
...
mkdir autograder/results/
#
# Docker run -ti -v `pwd`/autograder:/autograder -v `pwd`/sq/s10:/autograder/submission ubuntu-racket /bin/bash
docker run -ti \
		-v `pwd`/tests/sq/s10:/autograder/submission \
		-v `pwd`/autograder/results:/autograder/results \
		shriramk/gradescope-racket-sq /autograder/run_autograder
  C-c C-c^Cuser break
  context...:
   /usr/share/racket/collects/racket/contract/private/arrow-val-first.rkt:486:18
   /autograder/source/lib-grade.rkt:177:0: call/timeout
   /autograder/source/lib-grade.rkt:85:0: generate-results/hash
   /autograder/source/grade.rkt:18:0
   body of "/autograder/source/grade.rkt"
```

Commit:
🐛 fix: use racket/engine

Known issue, each test runs in sync, timeout is applied to each
individual test, not the entire suite :(

jasonhemann commented:
I still see a hang in the [above example](https://github.com/shriram/gradescope-racket/pull/23#issuecomment-1124973416). Investigating, maybe I can see something.

jasonhemann commented:
Nevermind, I just wasn't waiting long enough.

jasonhemann commented on lib-grade.rkt:
```diff
@@ -136,9 +136,20 @@
+  ;; (run-test-case/timeout name action (alarm-evt (+ (current-inexact-milliseconds) 2500))
+  (define (run-test-case/timeout name action)
+    (define eng (engine (lambda (b) (run-test-case name action))))
+    (if (engine-run (* 1000 (timeout-in-seconds)) eng)
```
Do we want to even construct the engine if `(timeout-in-seconds)` is `#f` ? I was thinking something like the following. Also so that we aren't, in case of a `#f` parameter value, multiplying by a boolean. 

```
    (if (timeout-in-seconds)
        (let ((eng (engine (lambda (p) (run-test-case name action)))))
          (if (engine-run (* 1000 (timeout-in-seconds)) eng)
              (engine-result eng)
              (test-error name (exn:fail "Execution timed out" (current-continuation-marks)))))
        (run-test-case name action))
``` 

A second question. Do timeouts deserve their own classification of things to report, on the same level as successes, failures, errors. Right now in the report we lose the timeout exception you raise, so it's not as informative as it could be.

rymaju replied:
Yeah, this is definitely not optimal. A different report for time outs would be better as well 🤔

jasonhemann reviewed (commented):
Thoughts

rymaju commented:
I'm going to try this again, I've been experimenting with futures but they seem to play poorly with engines.

rymaju commented:
So, I have a prototype here: https://gist.github.com/rymaju/ba70a1f2d025c2e873f099536686a3e2

Its hacky, but it works. I think the solution is right (engine running in a thread, whose result is placed into a channel which can be received by the main thread), but theres too many changes.

I'll listen to the Pragmatic Programmer for once and start from scratch again, im hoping to have a new PR that is (mostly) additive.
